app/src/main/java/org/nieghborhoodbikeworks/nbw/ui/queue/QueueFragment.java

```diff
@@ -1,114 +1,224 @@
-                mQueue = (ArrayList<String>) dataSnapshot.child("queue").getValue();
-                Log.d(TAG, "verifying");
+            public void onClick(View v) {
+                enqueueUser();
             }
+        });
 
+        mDequeueButton.setOnClickListener(new View.OnClickListener() {
             @Override
-            public void onCancelled(DatabaseError databaseError) {
-                Log.w(TAG, "loadQueue:onCancelled", databaseError.toException());
+            public void onClick(View v) {
+                dequeueUser();
             }
-        };
-        databaseReference.addValueEventListener(queueListener);
-
-        alertDialogBuilder = new AlertDialog.Builder(getActivity())
-                .setTitle("Queue")
-                .setMessage("Would you like to add yourself to the queue?")
-                .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
-                    public void onClick(DialogInterface dialog, int which) {
-//                      some logic for when the user decides to be added to the queue
-//                      databaseReference.child("queue").updateChildren();
-                      databaseReference.child("queue").setValue(user);
+        });
+
+    }
+
+    private void enqueueUser() {
+        // If there is no user currently signed in, mUser.getName() will throw a NPE
+        try {
+            Map<String, Object> childUpdates = new HashMap<>();
+            childUpdates.put(mUser.getName(), mUser);
+            mQueueDatabase.updateChildren(childUpdates);
+            updateQueue();
+            FirebaseAuth.getInstance().signOut();
+            mUser = null;
+        } catch (NullPointerException e) {
+            // Redirect the user to the login screen if they aren't signed in and would
+            // like to be added to, or removed from, the queue
+            Navigation.findNavController(view).navigate(R.id.loginFragment);
+        }
+    }
+
+    private void dequeueUser() {
+        // If there is no user currently signed in, mUser.getName() will throw a NPE
+        try {
+            mQueueDatabase.child(mUser.getName()).removeValue();
+            updateQueue();
+            FirebaseAuth.getInstance().signOut();
+            mUser = null;
+        } catch (NullPointerException e) {
+            // Redirect the user to the login screen if they aren't signed in and would
+            // like to be added to, or removed from, the queue
+            Navigation.findNavController(view).navigate(R.id.loginFragment);
+        }
+    }
+
+    /**
+     * The updateQueue method reads from the queue node in the database and adds the users to the mQueue
+     * ArrayList.
+     */
+    private void updateQueue() {
+        // Reads the data on the "queue" node in the database
+        ValueEventListener queueListener = new ValueEventListener() {
+            @Override
+            public void onDataChange(DataSnapshot dataSnapshot1) {
```

Comment:
Personal preference: have database methods in the SharedViewModel. lmk what you think.